// terra-api/lib/cjs/API/Data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetData = void 0;
const Helpers_1 = require("./Helpers");
function GetData(type, devId, apiKey, userId, startDate, endDate, toWebhook, withSamples, retryIfRateLimited) {
    const requestOptions = {
        method: 'GET',
        headers: {
            'X-API-Key': apiKey,
            'dev-id': devId,
            'Content-Type': 'application/json',
        },
    };
    let requestParams = {
        user_id: userId,
        start_date: (0, Helpers_1.DateToTerraDate)(startDate),
    };
    if (endDate) {
        requestParams = Object.assign({}, requestParams, {
            end_date: (0, Helpers_1.DateToTerraDate)(endDate),
        });
    }
    if (toWebhook !== undefined) {
        requestParams = Object.assign({}, requestParams, {
            to_webhook: toWebhook.toString(),
        });
    }
    if (withSamples !== undefined) {
        requestParams = Object.assign({}, requestParams, {
            with_samples: withSamples,
        });
    }
    if (retryIfRateLimited !== undefined) {
        requestParams = Object.assign({}, requestParams, {
            retry_if_rate_limited: retryIfRateLimited,
        });
    }
    return (0, Helpers_1.RequestWrapper)(`${type}`, requestOptions, requestParams);
}
exports.GetData = GetData;
