// terra-api/lib/cjs/API/UserInfo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeauthUser = exports.GetUser = void 0;
const cross_fetch_1 = __importDefault(require("cross-fetch"));
const Helpers_1 = require("./Helpers");
function GetUser(devId, apiKey, userID, referenceID) {
    const requestOptions = {
        method: 'GET',
        headers: {
            'X-API-Key': apiKey,
            'dev-id': devId,
            'Content-Type': 'application/json',
        },
    };
    return (0, Helpers_1.RequestWrapper)('userInfo', requestOptions, {
        user_id: userID,
        reference_id: referenceID,
    });
}
exports.GetUser = GetUser;
function DeauthUser(devId, apiKey, userId) {
    const requestOptions = {
        method: 'DELETE',
        headers: {
            'X-API-Key': apiKey,
            'dev-id': devId,
            'Content-Type': 'application/json',
        },
    };
    return new Promise((res, rej) => {
        (0, cross_fetch_1.default)(`https://api.tryterra.co/v2/auth/deauthenticateUser?user_id=${userId}`, requestOptions)
            .then((_) => res())
            .catch((error) => rej(error));
    });
}
exports.DeauthUser = DeauthUser;
