// terra-api/lib/esm/API/Data.js
import { DateToTerraDate, RequestWrapper } from './Helpers';
export function GetData(type, devId, apiKey, userId, startDate, endDate, toWebhook, withSamples, retryIfRateLimited) {
    const requestOptions = {
        method: 'GET',
        headers: {
            'X-API-Key': apiKey,
            'dev-id': devId,
            'Content-Type': 'application/json',
        },
    };
    let requestParams = {
        user_id: userId,
        start_date: DateToTerraDate(startDate),
    };
    if (endDate) {
        requestParams = Object.assign({}, requestParams, {
            end_date: DateToTerraDate(endDate),
        });
    }
    if (toWebhook !== undefined) {
        requestParams = Object.assign({}, requestParams, {
            to_webhook: toWebhook.toString(),
        });
    }
    if (withSamples !== undefined) {
        requestParams = Object.assign({}, requestParams, {
            with_samples: withSamples,
        });
    }
    if (retryIfRateLimited !== undefined) {
        requestParams = Object.assign({}, requestParams, {
            retry_if_rate_limited: retryIfRateLimited,
        });
    }
    return RequestWrapper(`${type}`, requestOptions, requestParams);
}
